src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java

```diff
@@ -1873,4 +1881,69 @@ private boolean expectedFunction(@Nullable PsiElement resolve) {
+    public ErlangFunctionCallParameterResolvingBySpec() {
+      super("functionCallParameterResolvingBySpec");
+    }
+
+    @Override
+    public boolean accepts(@NotNull T element, ProcessingContext context) {
+      ErlangQAtom argument = element instanceof ErlangQAtom ? (ErlangQAtom) element : null;
+      ErlangSpecification specification = getSpecification(argument);
+      if (argument == null || specification == null) return false;
+
+      int position = getPositionOfArgument(argument);
+      if (position == -1) return false;
+
+      ErlangFunTypeSigs signature = specification.getSignature();
+      if (signature == null) return false;
+
+      for (ErlangTypeSig typeSig : signature.getTypeSigList()) {
+        List<ErlangType> types = typeSig.getFunType().getFunTypeArguments().getTypeList();
+        if (types.size() <= position) continue;
+
+        ErlangQVar typeVar = types.get(position).getQVar();
+        String typeVarName = typeVar != null ? typeVar.getName() : null;
+        ErlangTypeSigGuard sigGuard = typeSig.getTypeSigGuard();
+        if (sigGuard == null || typeVarName == null) continue;
+
+        for (ErlangTypeGuard guard : sigGuard.getTypeGuardList()) {
+          ErlangTopType topType = PsiTreeUtil.getChildOfType(guard, ErlangTopType.class);
+          ErlangQVar var = PsiTreeUtil.getChildOfType(topType, ErlangQVar.class);
+          if (var != null && typeVarName.equals(var.getName())) {
+            for (ErlangType typeAndVar : guard.getTypeList()) {
+              ErlangType type = typeAndVar.getType();
+              if (type != null && "module()".equals(type.getText())) {
+                return true;
+              }
+            }
+          }
+        }
+      }
+      return false;
+    }
+
+    @Nullable
+    private static ErlangSpecification getSpecification(@Nullable ErlangQAtom argument) {
+      ErlangFunctionCallExpression call = getFunctionCallExpression(argument);
+      ErlangFunction function = resolveToFunction(call);
+      return function != null ? function.findSpecification() : null;
+    }
+
+    @Nullable
+    private static ErlangFunctionCallExpression getFunctionCallExpression(@Nullable ErlangQAtom atom) {
+      PsiElement maxExpression = atom != null ? atom.getParent() : null;
+      PsiElement argumentList = maxExpression != null ? maxExpression.getParent() : null;
+      return ObjectUtils.tryCast(argumentList != null ? argumentList.getParent() : null, ErlangFunctionCallExpression.class);
+     }
+
+    private static int getPositionOfArgument(@NotNull ErlangQAtom argument) {
+      ErlangMaxExpression maxExpression = ObjectUtils.tryCast(argument.getParent(), ErlangMaxExpression.class);
+      ErlangArgumentList argumentList = ObjectUtils.tryCast(maxExpression != null ? maxExpression.getParent() : null,
+                                                            ErlangArgumentList.class);
+      if (argumentList == null) return -1;
```

Comment:
I see, you're doing the same stuff here, as you've done in getFunctionCallExpression. Try traversing your PSI once.